sources/imdsclient/src/lib.rs

```diff
@@ -221,92 +241,109 @@ impl ImdsClient {
+                            .bytes()
+                            .await
+                            .context(error::ResponseBody {
+                                method: "GET",
+                                uri: &uri,
+                                code,
+                            })?
+                            .to_vec();
 
-                // IMDS returns 401 if the session token is expired or invalid
-                StatusCode::UNAUTHORIZED => {
-                    info!("Session token is invalid or expired");
-                    self.refresh_token().await?;
-                    info!("Refreshed session token");
-                    continue;
-                }
+                        let response_str = printable_string(&response_body);
 
-                StatusCode::REQUEST_TIMEOUT => {
-                    info!("Retrying request");
-                    continue;
-                }
+                        trace!("Response: {:?}", response_str);
 
-                code => {
-                    let response_body = response
-                        .bytes()
-                        .await
-                        .context(error::ResponseBody {
+                        error::Response {
                             method: "GET",
                             uri: &uri,
                             code,
-                        })?
-                        .to_vec();
-
-                    let response_str = printable_string(&response_body);
-
-                    trace!("Response: {:?}", response_str);
-
-                    return error::Response {
-                        method: "GET",
-                        uri: &uri,
-                        code,
-                        response_body: response_str,
+                            response_body: response_str,
+                        }
+                        .fail()
                     }
-                    .fail();
                 }
-            }
-        }
+            }),
+        )
+        .await
+        .context(error::TimeoutFetchIMDS)?
+    }
+
+    /// Fetches a new session token and writes it to the current ImdsClient.
+    async fn write_token(&self) -> Result<()> {
+        *self
+            .session_token
+            .write()
+            .map_err(|_| error::Error::FailedWriteToken {})? =
+            fetch_token(&self.client, &self.imds_base_uri, &self.retry_timeout).await?;
+        Ok(())
     }
 
-    /// Fetches a new session token and adds it to the current ImdsClient.
-    async fn refresh_token(&mut self) -> Result<()> {
-        self.session_token = fetch_token(&self.client, &self.imds_base_uri).await?;
-        Ok(())
+    /// Helper to read session token within the ImdsClient
+    async fn read_token(&self) -> Result<Option<String>> {
+        let session_token = match self
+            .session_token
+            .read()
+            .map_err(|_| error::Error::FailedReadToken {})?
+            .clone()
```

Comment:
nit: It is a bit unusual to return a `clone()` of a value rather than a reference. It might be worth leaving a comment that the idea is to allow the `RwLock` to be released as soon as possible, as this detail might not be so obvious as time passes.